src/dbus_api/blockdev.rs

```diff
@@ -0,0 +1,285 @@
+                        MethodErr::failed(&format!("no path for parent object path {}",
+                                                   &blockdev_data.parent))
+                    })?;
+
+    let pool_uuid = pool_path
+        .get_data()
+        .as_ref()
+        .ok_or_else(|| MethodErr::failed(&format!("no data for object path {}", object_path)))?
+        .uuid;
+
+    let engine = dbus_context.engine.borrow();
+    let pool =
+        engine
+            .get_pool(&pool_uuid)
+            .ok_or_else(|| {
+                            MethodErr::failed(&format!("no pool corresponding to uuid {}",
+                                                       &pool_uuid))
+                        })?;
+    let blockdev_uuid = &blockdev_data.uuid;
+    let blockdev =
+        pool.get_blockdev(blockdev_uuid)
+            .ok_or_else(|| MethodErr::failed(&format!("noblockdev with uuid {}", &blockdev_uuid)))?;
+    i.append(getter(blockdev)?);
+    Ok(())
+}
+
+/// Get the devnode for an object path.
+fn get_blockdev_devnode(i: &mut IterAppend,
+                        p: &PropInfo<MTFn<TData>, TData>)
+                        -> Result<(), MethodErr> {
+    get_blockdev_property(i,
+                          p,
+                          |p| Ok(MessageItem::Str(format!("{}", p.devnode().display()))))
+}
+
+fn get_blockdev_hardware_id(i: &mut IterAppend,
+                            p: &PropInfo<MTFn<TData>, TData>)
+                            -> Result<(), MethodErr> {
+    fn get_hardware_id(blockdev: &BlockDev) -> Result<MessageItem, MethodErr> {
+        match *blockdev.hardware_id() {
+            Some(ref id) => Ok(MessageItem::Str(id.to_owned())),
+            None => Ok(MessageItem::Str("".to_owned())),
+        }
+    }
+
+    get_blockdev_property(i, p, get_hardware_id)
+}
+
+fn get_blockdev_user_id(i: &mut IterAppend,
+                        p: &PropInfo<MTFn<TData>, TData>)
+                        -> Result<(), MethodErr> {
+    fn get_user_id(blockdev: &BlockDev) -> Result<MessageItem, MethodErr> {
+        match *blockdev.user_id() {
+            Some(ref id) => Ok(MessageItem::Str(id.to_owned())),
+            None => Ok(MessageItem::Str("".to_owned())),
+        }
+    }
+
+    get_blockdev_property(i, p, get_user_id)
+}
+
+fn get_blockdev_initialization_time(i: &mut IterAppend,
+                                    p: &PropInfo<MTFn<TData>, TData>)
+                                    -> Result<(), MethodErr> {
+    get_blockdev_property(i,
+                          p,
+                          |p| Ok(MessageItem::Str(p.initialization_time().to_rfc3339())))
+}
+
+fn get_blockdev_physical_size(i: &mut IterAppend,
+                              p: &PropInfo<MTFn<TData>, TData>)
+                              -> Result<(), MethodErr> {
+    get_blockdev_property(i,
+                          p,
+                          |p| Ok(MessageItem::Str(format!("{}", *p.total_size()))))
+}
+
+fn get_blockdev_state(i: &mut IterAppend,
```

Comment:
This looks like it might be done better if done in the same way as the error values and redundancy values, i.e., codes are exported as part of the Manager API and the state itself is exported as a number. The most obvious example is Redundancy.

Defining BlockDevState and exporting codes could be put in a separate PR, for quicker merging.

Reply:
I think this comment is superseded by the discussions we've had more recently about removing the other exported codes, and relying on the API spec.

Reply:
That's correct. The method looks good to me.